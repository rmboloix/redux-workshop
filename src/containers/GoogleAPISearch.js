import React, {Component} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {googleAPISearch} from '../actions/index';

class GoogleAPISearch extends Component {

  constructor(props) {
    super(props);

    this.state = {
      term: ''
    };

    this.changeFilterText = this.changeFilterText.bind(this);
    this.doSearch = this.doSearch.bind(this);
  }

  changeFilterText(event) {
    this.setState({
      term: event.target.value
    })
  }

  doSearch(event) {
    event.preventDefault();
    this.props.googleAPISearch(this.state.term);
  }

  render() {
    return (
      <form onSubmit={this.doSearch}>
        <div className="input-group">
          <input type="text"
                 className="form-control"
                 value={this.state.term}
                 onChange={this.changeFilterText}
                 placeholder="Clean code"/>
          <span className="input-group-btn">
            <input className="btn btn-primary"
                    type="submit"
                    value="Search on Google" />
          </span>
        </div>
      </form>
    );
  }
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({googleAPISearch}, dispatch);
}

export default connect(null, mapDispatchToProps)(GoogleAPISearch);
