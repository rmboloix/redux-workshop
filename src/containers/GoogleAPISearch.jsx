import React, {Component} from 'react'
import {connect} from 'react-redux'

import {newSearch} from '../actions'
import {searchTypes} from '../api/GoogleBooks'

class GoogleAPISearch extends Component {
  constructor (props) {
    super(props)

    this.state = {
      term: '',
      type: 'title',
      maxResults: 10,
      placeHolder: 'Clean Code'
    }

    this.handleInputChange = this.handleInputChange.bind(this)
    this.handleTypeSelectorChange = this.handleTypeSelectorChange.bind(this)
    this.handleMaxResultsChange = this.handleMaxResultsChange.bind(this)
    this.doSearch = this.doSearch.bind(this)
  }

  handleInputChange (event) {
    this.setState({
      term: event.target.value
    })
  }

  handleTypeSelectorChange (event) {
    let placeHolder = ''
    let value = event.target.value

    switch (value) {
      case 'title' :
        placeHolder = 'Clean Code'
        break
      case 'author':
        placeHolder = 'Robert C. Martin'
        break
      case 'publisher' :
        placeHolder = "O'Reilly Media, Inc"
        break
      case 'isbn':
        placeHolder = '9781785288319'
        break
    }

    this.setState({
      type: value,
      placeHolder
    })
  }

  handleMaxResultsChange (event) {
    this.setState({
      maxResults: event.target.value
    })
  }

  doSearch (event) {
    event.preventDefault()

    let criteria = {
      [this.state.type]: this.state.term.trim(),
      [searchTypes.subject]: 'Computers'
    }

    this.props.newSearch(criteria, this.state.maxResults)
  }

  render () {
    return (
      <form onSubmit={this.doSearch} className='row search-form'>
        <div className='col-md-2 col-sm-2'>
          <select className='form-control'
            onChange={this.handleTypeSelectorChange}>
            <option value={searchTypes.title}>Title</option>
            <option value={searchTypes.author}>Author</option>
            <option value={searchTypes.publisher}>Publisher</option>
            <option value={searchTypes.raw}>Raw</option>
            <option value={searchTypes.isbn}>ISBN</option>
          </select>
        </div>
        <div className='col-md-1 col-sm-1'>
          <select className='form-control'
            onChange={this.handleMaxResultsChange}>
            <option value={10}>10</option>
            <option value={20}>20</option>
            <option value={40}>40</option>
          </select>
        </div>
        <div className='col-md-7 col-sm-7'>
          <input type='text'
            className='form-control'
            value={this.state.term}
            onChange={this.handleInputChange}
            placeholder={this.state.placeHolder} />
        </div>
        <div className='col-md-2 col-sm-2'>
          <button className='btn form-control'
            type='submit'>
            {this.props.searchStatus !== 'SEARCHING' ? 'Search on Google' : (<i className='fa fa-spinner fa-spin' />)}
          </button>
        </div>
      </form>
    )
  }
}

function mapStatusToProps (state) {
  return {
    searchStatus: state.search.status
  }
}

export default connect(mapStatusToProps, {newSearch})(GoogleAPISearch)
