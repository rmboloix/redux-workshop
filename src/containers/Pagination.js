import React, {Component} from 'react';
import {connect} from 'react-redux';
import {changePage} from '../actions/index';

/**
 * Created by rafael on 27/03/17.
 */
class Pagination extends Component {
  constructor(props) {
    super(props);

    this.changePage = this.changePage.bind(this);
  }

  changePage(newPage){
    this.props.changePage(newPage);
  }

  render() {

    let {activePage, totalPages} = this.props.search;


    //TODO change selector edges cases behaviour...
    let previousPageSelector = activePage !== 1 && (<li>
        <a href="#" aria-label="Previous" onClick={() => {
          this.changePage(activePage - 1)
        }}>
          <span aria-hidden="true">&laquo; Prev</span>
        </a>
      </li>);

    let nextPageSelector = totalPages > activePage &&(<li>
        <a href="#" aria-label="Next" onClick={() => {
          this.changePage(activePage + 1)
        }}>
          <span aria-hidden="true">Next &raquo;</span>
        </a>
      </li>);

    return (
      this.props.search.status === 'DONE' &&
      <nav aria-label="Search navigation">
        <ul className="pagination">
          {previousPageSelector}
          {/*{pages}*/}
          {nextPageSelector}
        </ul>
      </nav>
    );
  }
}

function mapStateToProps(state) {
  return {
    search: state.search
  }
}

export default connect(mapStateToProps, {changePage})(Pagination);