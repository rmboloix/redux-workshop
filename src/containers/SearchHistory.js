import React, {Component} from 'react';
import {connect} from 'react-redux';
import {newSearch} from './../actions/index';

export class SearchHistory extends Component {

  constructor(props) {
    super(props);

    this.handleSearch = this.handleSearch.bind(this);
  }

  handleSearch(search) {
    this.props.newSearch(
      search.searchTerm,
      search.searchType,
      10,
      0
    );
  }


  render() {
    let searchList = this.props.searchHistory.map((search, index) => {
      return (
        <button
          key={index}
          type="button"
          className="list-group-item"
          onClick={() => {this.handleSearch(search)}}
        >
          {search.searchTerm} <span className="badge">{search.searchType}</span>
        </button>
      );
    });

    return (
      <div className="list-group">
        {searchList}
      </div>
    );
  }
}

function mapStateToProps(state) {
  return {
    searchHistory: state.searchHistory
  }
}

export default connect(mapStateToProps, {newSearch})(SearchHistory);