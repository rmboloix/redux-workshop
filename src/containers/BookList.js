import React, {Component} from 'react';
import {connect} from 'react-redux';
import Book from '../components/book/Book';
import BookContainer from '../components/book/BookContainer';
import Pagination from './Pagination';
import TotalResults from '../components/TotalResults';

class BookList extends Component {

  constructor(props) {
    super(props);
  }

  render() {
    let activeClass = this.props.activeView === '3books-view'
      ? 'col-md-4 col-sm-6' : 'col-md-12 col-sm-12';

    let books = this.props.books.map((book, index) => {
      return (
        <BookContainer key={book.id}
                       activeClass={activeClass}>
          <Book book={book}/>
        </BookContainer>
      );
    });

    let pagination = this.props.totalItems > 0 &&
      (<div className="col-md-12 text-center">
        <Pagination />
      </div>);

    return (
      <div className="row">
        <div className="col-md-2">
          <TotalResults totalItems={this.props.totalItems}/>
        </div>
        <div className="col-md-12">
          {books}
        </div>
        {pagination}
      </div>

    )
  }
}

function mapStateToProps(state) {
  return {
    books: state.books,
    totalItems: state.search.totalItems,
    activeView: state.activeView
  }
}

export default connect(mapStateToProps)(BookList);
