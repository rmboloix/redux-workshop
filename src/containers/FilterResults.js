import React from 'react';
import { connect } from 'react-redux';
import SearchHistory from './SearchHistory';
import ViewSelector from "./ViewSelector";

const FilterResults = (props) => {
  return (
    <div>
      <div className="panel-filter panel panel-default">
        <div className="panel-heading">
            <ViewSelector />
        </div>
        <div className="panel-body">
          Total results <span className="label label-primary">{props.totalItems}</span>
        </div>
      </div>
      <SearchHistory />
    </div>
  );
};

function mapStateToProps(state) {
  return {
    totalItems: state.search.totalItems
  };
}

export default connect(mapStateToProps)(FilterResults);