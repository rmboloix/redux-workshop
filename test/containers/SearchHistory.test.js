import React from 'react';
import renderer from 'react-test-renderer';
import {SearchHistory} from './../../src/containers/SearchHistory';

const history = [
    {
      searchTerm: 'docker',
      searchType: 'title'
    },
    {
      searchTerm: 'kubernetes',
      searchType: 'title'
    },
    {
      searchTerm: 'robert c. martin',
      searchType: 'author'
    }
  ],
  newSearch = jest.fn();

describe('SearchHistory component', () => {
  it('should render empty list', () => {

    let tree = renderer.create(
      <SearchHistory
        searchHistory={[]}
        newSearch={newSearch}/>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });


  it('should render history list', () => {
    let tree = renderer.create(
      <SearchHistory
        searchHistory={history}
        newSearch={newSearch}/>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });


  it('should do a new search when button is clicked', () => {
    let tree = renderer.create(
      <SearchHistory
        searchHistory={history}
        newSearch={newSearch}/>
    ).toJSON();

    tree.children[2].props.onClick();

    expect(newSearch.mock.calls[0])
      .toEqual(['robert c. martin', 'author', 10, 0]);
  });
});