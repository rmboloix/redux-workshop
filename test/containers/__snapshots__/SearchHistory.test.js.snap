// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchHistory component should render empty list 1`] = `
<div
  className="panel-filter panel panel-default"
>
  <div
    className="panel-heading"
  >
    <h3
      className="panel-title"
    >
      Search History
    </h3>
  </div>
  <div
    className="panel-body"
  >
    <div
      className="list-group"
    />
  </div>
</div>
`;

exports[`SearchHistory component should render history list 1`] = `
<div
  className="panel-filter panel panel-default"
>
  <div
    className="panel-heading"
  >
    <h3
      className="panel-title"
    >
      Search History
    </h3>
  </div>
  <div
    className="panel-body"
  >
    <div
      className="list-group"
    >
      <button
        className="list-group-item"
        onClick={[Function]}
        type="button"
      >
        docker
         
        <span
          className="badge"
        >
          title
        </span>
      </button>
      <button
        className="list-group-item"
        onClick={[Function]}
        type="button"
      >
        kubernetes
         
        <span
          className="badge"
        >
          title
        </span>
      </button>
      <button
        className="list-group-item"
        onClick={[Function]}
        type="button"
      >
        robert c. martin
         
        <span
          className="badge"
        >
          author
        </span>
      </button>
    </div>
  </div>
</div>
`;
